### PR TITLE
No longer crash 64 bit Adobe Acrobat / Reader when rendering a virtualBuffer for PDF content

### LINKED ISSUE #12920: NVDA does not work with Adobe reader DC 64 bit
### Steps to reproduce:
1. download and install latest Adobe reader 64 bit
2. open a document in Adobe Reader
### Actual behavior:
NVDA plays multiple error tones, and Adobe reader exits with no visible error. The following appears  in log
ERROR - RPC process 10084 (Acrobat.exe) (14:56:39.824) - Dummy-1229 (4232): 
Thread 8124, build\x86_64\remote\COMProxyRegistration.cpp, registerCOMProxy, 182: 
Unable to register interface IAccessibleHyperlink with proxy stub IAccessible2Proxy.dll, code -2147023156 
### Expected behavior:
Adobe reader does not exit and NVDA reads document.
#### NVDA installed/portable/running from source:
installed
#### NVDA version:
alpha-23897,7bd9e8b1
#### Windows version:
windows 10 21h1
#### Name and version of other software in use when reproducing the issue:
Adobe reader DC
#### Does the issue still occur after restarting your computer?
yes
#### If NVDA add-ons are disabled, is your problem still occurring?
yes
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
yes

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
Fixes #12920

### Summary of the issue:
Adobe Acrobat / Reader 64 bit crashes when NVDA renders a virtualBuffer for PDF content.
There are two reasons for this:
1. Adobe's IAccID custom interface has been changed so that specifically on 64 bit builds, the ID argument is now 64 bit wide instead of 32 bit. This meant that NVDA was using the wrong data type and wrong call signature for IAccID::get_accID, and therefore Acrobat / Reader was trying to write a 64 bit value in to the space that could only hold 32 bits, causing a buffer overrun.
2. Adobe Acrobat / Reader's accID values, including the MSAA child IDs used for IAccessible::get_accChild and AccessibleObjectFromEvent were 64 bit wide, though MSAA only supports 32 bit values. This would cause invalid and truncated values to be used for child IDs, in many cases causing a crash.

Adobe has addressed point 2 in Adobe Acrobat / Reader beta versions and will be in a stable build in the near future.
But NVDA needs to address point 1 in its own code.

### Description of user facing changes
Using a version of Adobe Acrobat / Reader with the required fixes on adobe's side, along with this PR in NVDA, the crash is avoided.
 
### Description of development approach
miscDeps has been updated to contain the newest version of acrobatAccess.idl from the Adobe Acrobat SDK.
The adobeAcrobat vbufBackend getAccID method has been changed to fetch the out param of IAccID::get_accID with a LONG_PRT which will be 64 bit wide on 64 bit Acrobat / Reader, and 32 bit wide on 32 bit Acrobat / Reader. The value is then static casted to long when returned. As Acrobat always uses 32 bit accID vlaues, no information will be lost in the conversion. 

### Testing strategy:
Opened multiple pdf documents in Adobe Acrobat / Reader while running NVDA. Ensured that Acrobat / Reader did not crash and that the pdf could be read with the arrow keys.

### Known issues with pull request:
None known.

### Change log entries:
New features
Changes
Bug fixes
* Adobe Acrobat / Reader 64 bit will no longer crash when reading a PDF document. (#12920)
  * Note that the most up to date version of Adobe Acrobat / Reader is also required to avoid the crash.
For Developers

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

### CODE REVIEW

**Comment by michaelDCurran** [comment]
Although this requires an Acrobat / Reader with fixes to fully avoid the crash, this can still be merged to master safely. It however will not make the crash better until the fixes is available on the adobe side also.

**Review by LeonarddeR — APPROVED** [review]

**Review by OzancanKaratas — APPROVED** [review]

File: nvdaHelper/vbufBackends/adobeAcrobat/adobeAcrobat.cpp [hunk 1]
```diff
@@ -56,16 +55,20 @@ long getAccID(IServiceProvider* servprov) {
 	} 
 	LOG_DEBUG(L"IAccID at "<<paccID);
 
+	// IAccID::get_accID takes a longlong on 64 bit and a long on 32 bit.
+	// However, Acrobat will internally only place a 32 bit value into ID.
+	// Thus we call it with a LONG_PTR* and can safely static cast it to a long.
+	LONG_PTR ID = 0;
 	LOG_DEBUG(L"Calling get_accID");
-	if((res=paccID->get_accID((long*)(&ID)))!=S_OK) {
+	if((res=paccID->get_accID(&ID))!=S_OK) {
```
Comment (seanbudd) [inline nvdaHelper/vbufBackends/adobeAcrobat/adobeAcrobat.cpp:63]: Minor linting fix for readability

```suggestion
	if ((res = paccID->get_accID(&ID)) != S_OK) {
```

**Review by seanbudd — APPROVED** [review]

**Comment by seanbudd** [comment]
@CyrilleB79 - it is safe to merge, but without an Adobe build to test with, there is no way to be certain that the issue is fixed.
There is also no urgency to merge this without a beta/public Adobe build, at least until we branch for 2022.3.
I have added another label to reflect the fact that this will go into 2022.3.

**Comment by OzancanKaratas** [comment]
Adobe has already released a bug fix. Please check for updates in your product. I tested it and this pull request works as expected. See the release notes [here](https://www.adobe.com/devnet-docs/acrobatetk/tools/ReleaseNotesDC/continuous/dccontinuousjune2022.html#dccontinuousjuntwentytwentytwo).

**Comment by seanbudd** [comment]
@OzancanKaratas I have no trouble using that build of Adobe Reader with NVDA 2022.2beta2,
i.e. #12920 is already fixed on Adobe's end for me using stable NVDA.
I believe the issue that this PR fixes is waiting on a future Adobe build, I will need to confirm.

**Comment by seanbudd** [comment]
@OzancanKaratas - do you have the same experience as me?
Are you, or anyone else, still experiencing the crash using the latest Adobe reader and the latest beta/stable/alpha NVDA?

**Comment by k-kolev1985** [comment]
I can still reproduce this.

Test environment:
- Operating system: Windows 11 Pro version 21H2 (build 22000.739), 64-bit, in Bulgarian with all locale settings set to "Bulgarian".
- NVDA version: alpha-25815,8a237d40, installed.
- Adobe Acrobat DC version: 2022.001.20142, 64-bit, in english.

**Comment by seanbudd** [comment]
Thanks @k-kolev1985: does [this PR build](https://ci.appveyor.com/api/buildjobs/vbsks73sujr13u26/artifacts/output%2Fnvda_snapshot_pr13852-25717%2Cf4ff2aa4.exe) fix the issue?

**Comment by k-kolev1985** [comment]
Yes, it does.

File: nvdaHelper/vbufBackends/adobeAcrobat/adobeAcrobat.cpp [hunk 2]
```diff
@@ -56,16 +55,20 @@ long getAccID(IServiceProvider* servprov) {
...
+	if ((res = paccID->get_accID(&ID)) != S_OK) {
 		LOG_DEBUG(L"paccID->get_accID returned "<<res);
 		ID = 0;
 	}
 
 	LOG_DEBUG("Releasing IAccID");
 	paccID->Release();
 
-	return ID;
+	return static_cast<long>(ID);
```
[the remark this answers is not in the document]
Comment (feerrenrut) [reply nvdaHelper/vbufBackends/adobeAcrobat/adobeAcrobat.cpp:71]: Perhaps a check for this (and error log if not) should be added. That is, check if the value is a safe 32bit value.

**Pull request merged**